service/matching/version_sets.go

```diff
@@ -0,0 +1,196 @@
+func UpdateVersionSets(existingData *persistence.VersioningData, req *workflowservice.UpdateWorkerBuildIdCompatabilityRequest, maxSize int) error {
+	err := updateImpl(existingData, req)
+	if err != nil {
+		return err
+	}
+	// Limit graph size if it's grown too large
+	depthLimiter(existingData, maxSize, true)
+	return nil
+}
+
+func updateImpl(existingData *persistence.VersioningData, req *workflowservice.UpdateWorkerBuildIdCompatabilityRequest) error {
+	// First find if the targeted version is already in the sets
+	targetedVersion := extractTargetedVersion(req)
+	targetSetIx, versionInSetIx := findVersion(existingData, targetedVersion)
+
+	if _, ok := req.GetOperation().(*workflowservice.UpdateWorkerBuildIdCompatabilityRequest_AddNewVersionIdInNewDefaultSet); ok {
+		// If it's not already in the sets, add it as the new default set
+		if targetSetIx != -1 {
+			return serviceerror.NewInvalidArgument(fmt.Sprintf("version %s already exists", targetedVersion))
+		}
+
+		existingData.VersionSets = append(existingData.GetVersionSets(), &taskqueuepb.CompatibleVersionSet{
+			Versions: []string{targetedVersion},
+		})
+	} else if op, ok := req.GetOperation().(*workflowservice.UpdateWorkerBuildIdCompatabilityRequest_AddNewCompatibleVersion_); ok {
+		compatVer := op.AddNewCompatibleVersion.GetExistingCompatibleVersion()
+		compatSetIx, _ := findVersion(existingData, compatVer)
+		if compatSetIx == -1 {
+			return serviceerror.NewNotFound(
+				fmt.Sprintf("targeted compatible_version %v not found", compatVer))
+		}
+		if targetSetIx != -1 {
+			// If the version does exist, this operation can't do anything meaningful, but we can fail if the user
+			// says the version is now compatible with some different set.
+			return serviceerror.NewInvalidArgument(fmt.Sprintf("version %s already exists", targetedVersion))
+		}
+
+		// If the version doesn't exist, add it to the compatible set
+		existingData.VersionSets[compatSetIx].Versions =
+			append(existingData.VersionSets[compatSetIx].Versions, targetedVersion)
+		if op.AddNewCompatibleVersion.GetMakeSetDefault() {
+			makeDefaultSet(existingData, compatSetIx)
+		}
+	} else if _, ok := req.GetOperation().(*workflowservice.UpdateWorkerBuildIdCompatabilityRequest_PromoteSetByVersionId); ok {
+		if targetSetIx == -1 {
+			return serviceerror.NewNotFound(fmt.Sprintf("targeted version %v not found", targetedVersion))
+		}
+		makeDefaultSet(existingData, targetSetIx)
+	} else if _, ok := req.GetOperation().(*workflowservice.UpdateWorkerBuildIdCompatabilityRequest_PromoteVersionIdWithinSet); ok {
+		if targetSetIx == -1 {
+			return serviceerror.NewNotFound(fmt.Sprintf("targeted version %v not found", targetedVersion))
+		}
+		makeVersionInSetDefault(existingData, targetSetIx, versionInSetIx)
+	}
+
+	return nil
+}
+
+func extractTargetedVersion(req *workflowservice.UpdateWorkerBuildIdCompatabilityRequest) string {
+	if req.GetAddNewCompatibleVersion() != nil {
```

Comment:
kinda weird to effectively do this type switch twice, and using a different technique too. I know go (nor proto) doesn't have the pattern matching that would make this nice, but maybe better to do one switch and just pull these out of the typed value?

Reply:
Problem is I also need to get it out first to pass it into `findVersion`. I've made the styles consistent now but I don't immediately see a non-super-duplicative way to get rid of this fn.

Reply:
yeah, I meant duplicate calls to findVersion in each branch. it's more duplication in some ways (code in branches of type switch) and less in others (type switch itself). overall I'm guessing it comes out shorter. but it's subjective so :shrug: